server/mailet/rate-limiter/src/main/scala/org/apache/james/transport/mailets/PerRecipientRateLimitMailet.scala

```diff
@@ -0,0 +1,133 @@
+    val keyPrefix: Option[KeyPrefix] = Option(getInitParameter("keyPrefix")).map(KeyPrefix)
+    exceededProcessor = getInitParameter("exceededProcessor", Mail.ERROR)
+
+    val countRateLimiter: Option[PerRecipientRateLimiter] = Option(getInitParameter("count"))
+      .flatMap(parameter => extractCountToRateLimiter(parameter, duration, rateLimiterFactory)
+        .map(rateLimiter => PerRecipientRateLimiter(rateLimiter, keyPrefix, Count)))
+
+    val sizeRateLimiter: Option[PerRecipientRateLimiter] = Option(getInitParameter("size"))
+      .map {
+        case "" => throw new IllegalArgumentException("'size' field cannot be empty if specified")
+        case s => s
+      }
+      .flatMap(parameter => extractSizeToRateLimiter(parameter, duration, rateLimiterFactory)
+        .map(rateLimiter => PerRecipientRateLimiter(rateLimiter, keyPrefix, Size)))
+
+    rateLimiters = Seq(countRateLimiter, sizeRateLimiter)
+      .filter(limiter => limiter.isDefined)
+      .map(limiter => limiter.get)
+  }
+
+  override def service(mail: Mail): Unit = {
+    if (!mail.getRecipients.isEmpty) {
+      val rateLimitResults: Seq[(MailAddress, RateLimitingResult)] = applyRateLimiter(mail)
+
+      val rateLimitedRecipients: Seq[MailAddress] = rateLimitResults.filter(_._2.equals(RateExceeded)).map(_._1)
+      val acceptableRecipients: Seq[MailAddress] = rateLimitResults.filter(_._2.equals(AcceptableRate)).map(_._1)
+
+      (acceptableRecipients, rateLimitedRecipients) match {
+        case (acceptable, _) if acceptable.isEmpty => mail.setState(exceededProcessor)
+        case (_, exceeded) if exceeded.isEmpty => // do nothing
+        case _ =>
+          mail.setRecipients(ImmutableList.copyOf(acceptableRecipients.asJava))
+
+          Using(mail.duplicate())(newMail => {
+            newMail.setRecipients(ImmutableList.copyOf(rateLimitedRecipients.asJava))
+            getMailetContext.sendMail(newMail, exceededProcessor)
+          })(LifecycleUtil.dispose(_))
+      }
+    }
+  }
+
+  @VisibleForTesting
+  def parseDuration(): Duration =
+    Option(getInitParameter("duration"))
+      .map(string => DurationParser.parse(string, ChronoUnit.SECONDS))
+      .getOrElse(throw new IllegalArgumentException("'duration' is compulsory"))
+
+
+  private def extractCountToRateLimiter(quantity: String, duration: Duration, rateLimiterFactory: RateLimiterFactory): Option[RateLimiter] =
+    Some(quantity)
+      .map(_.toLong)
+      .map(AllowedQuantity.liftOrThrow)
+      .map(quantity => Rules(Seq(Rule(quantity, duration))))
+      .map(rateLimiterFactory.withSpecification)
+
+  private def extractSizeToRateLimiter(quantity: String, duration: Duration, rateLimiterFactory: RateLimiterFactory): Option[RateLimiter] =
+    Some(quantity)
+      .map(org.apache.james.util.Size.parse)
+      .map(_.asBytes())
+      .map(AllowedQuantity.liftOrThrow)
+      .map(quantity => Rules(Seq(Rule(quantity, duration))))
+      .map(rateLimiterFactory.withSpecification)
```

Comment:
You should be able to reuse much of the code abstracted by Quan to do just this.

Please rebase and refactor...